### PR TITLE
mds: MDLog::_recovery_thread: handle the errors gracefully

### PR DESCRIPTION
**Pull request by joscollin**
Fixes: https://tracker.ceph.com/issues/61201

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by joscollin** [comment]
jenkins test make check

File: src/mds/MDLog.cc [hunk 1]
```diff
@@ -977,8 +977,13 @@ void MDLog::_recovery_thread(MDSContext *completion)
     inodeno_t const default_log_ino = MDS_INO_LOG_OFFSET + mds->get_nodeid();
     jp.front = default_log_ino;
     int write_result = jp.save(mds->objecter);
-    // Nothing graceful we can do for this
-    ceph_assert(write_result >= 0);
+    if (write_result < 0) {
+      if (!mds_is_shutting_down) {
+	mds->damaged_unlocked();
+	ceph_abort();  // damaged should never return
```
Comment (dparmar18) [inline src/mds/MDLog.cc:983]:
```suggestion
	ceph_abort_msg("mds." + mds->get_nodeid() + " is damaged");  // damaged should never return
```

Comment (joscollin) [reply src/mds/MDLog.cc:983]: It doesn't make any difference.

**Review by gregsfortytwo — APPROVED** [review]
LGTM

**Review by gregsfortytwo — COMMENTED** [review]
This looks okay to me at a glance, but a consequence of this is that when shutting down we will fall through the if-block and keep running the rest of the function — which can try and do all kinds of things like erase journal blocks. That seems like it has the potential to cause further problems — probably we should just return or shut down the thread instead?

**Review by vshankar — APPROVED** [review]
Nice work!

**Comment by joscollin** [comment]
@vshankar @gregsfortytwo addressed the review comments. Using `is_daemon_stopping()` seems to be straight forward.

**Comment by joscollin** [comment]
jenkins test api

**Comment by joscollin** [comment]
jenkins test api

**Comment by joscollin** [comment]
jenkins test api

**Review by vshankar — APPROVED** [review]

**Comment by lxbsz** [comment]
@joscollin Could we add a detail comment in the commit to explain why we do this fix. Thanks.

**Comment by joscollin** [comment]
> @joscollin Could we add a detail comment in the commit to explain why we do this fix. Thanks.

Updated

File: src/mds/MDLog.cc [hunk 2]
```diff
@@ -977,8 +977,13 @@ void MDLog::_recovery_thread(MDSContext *completion)
     inodeno_t const default_log_ino = MDS_INO_LOG_OFFSET + mds->get_nodeid();
     jp.front = default_log_ino;
     int write_result = jp.save(mds->objecter);
-    // Nothing graceful we can do for this
-    ceph_assert(write_result >= 0);
+    if (write_result < 0) {
+      if (mds->is_daemon_stopping()) {
```
Comment (lxbsz) [inline src/mds/MDLog.cc:981]: We need to take the `mds_lock` before this ?

Comment (vshankar) [reply src/mds/MDLog.cc:981]: You are right -- the lock ins't taken right at the stat but at places.

Comment (joscollin) [reply src/mds/MDLog.cc:981]: okay, then the code should change from `mds->damaged_unlocked()` to `mds->damaged()`.

Comment (joscollin) [reply src/mds/MDLog.cc:981]: Because `mds->damaged_unlocked()` already has a lock

Comment (joscollin) [reply src/mds/MDLog.cc:982]: updated

**Comment by joscollin** [comment]
jenkins test make check

**Comment by joscollin** [comment]
jenkins test docs

**Comment by joscollin** [comment]
rebased

**Comment by joscollin** [comment]
jenkins test docs

**Comment by vshankar** [comment]
https://pulpito.ceph.com/?branch=wip-vshankar-testing-20230704.040242

**Comment by joscollin** [comment]
> https://pulpito.ceph.com/?branch=wip-vshankar-testing-20230704.040242

Looks good.

**Comment by vshankar** [comment]
jenkins test docs

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#13-July-2023

**Comment by vshankar** [comment]
@joscollin What's with the failing ceph-pr-docs test?

**Comment by joscollin** [comment]
jenkins test docs

**Comment by vshankar** [comment]
> jenkins test docs

That does not help since it triggers docs build check not sphinx build which seems different.

**Comment by joscollin** [comment]
> > jenkins test docs
> 
> That does not help since it triggers docs build check not sphinx build which seems different.

I gave a rerun, but it doesn't trigger a re-run.

**Comment by joscollin** [comment]
jenkins test docs.*

**Comment by joscollin** [comment]
> jenkins test docs.*

That^ didn't work. https://github.com/ceph/ceph-build/blob/main/ceph-pr-docs/config/definitions/ceph-pr-docs.yml

**Review by kotreshhr — APPROVED** [review]

**Comment by vshankar** [comment]
@joscollin maybe just push an update?

**Comment by joscollin** [comment]
rebased again. Let's see how this goes.

**Comment by joscollin** [comment]
jenkins test windows

**Comment by vshankar** [comment]
jenkins test windows

**Comment by joscollin** [comment]
> jenkins test windows

All PRs fail with the same reason `sed: can't read ./tools/build/src/tools/gcc.py: No such file or directory`

**Comment by vshankar** [comment]
jenkins test windows

**Comment by joscollin** [comment]
jenkins test windows

**Comment by vshankar** [comment]
@petrutlucian94 Would you know about the windows jenkins failures that have started showing up lately - https://github.com/ceph/ceph/pull/52173#issuecomment-1635689535

**Comment by petrutlucian94** [comment]
> @petrutlucian94 Would you know about the windows jenkins failures that have started showing up lately - [#52173 (comment)](https://github.com/ceph/ceph/pull/52173#issuecomment-1635689535)

Yes, bumping the Boost version caused a few problems. It's fixed here, should merge soon: https://github.com/ceph/ceph/pull/52427.

Thanks for bringing it up!

**Comment by vshankar** [comment]
> > @petrutlucian94 Would you know about the windows jenkins failures that have started showing up lately - [#52173 (comment)](https://github.com/ceph/ceph/pull/52173#issuecomment-1635689535)
> 
> Yes, bumping the Boost version caused a few problems. It's fixed here, should merge soon: #52427.
> 
> Thanks for bringing it up!

Ah :/

So, I will merge this PR since the issue is unrealted.

**Pull request merged**